File: docs/src/test/scala/docs/home/journal/EventTitleColumn.scala
/*
 * Copyright (C) 2022 - 2025 Lightbend Inc. <https://akka.io>
 */

package docs.home.journal

// #additional-column
import akka.persistence.r2dbc.journal.scaladsl.AdditionalColumn

// #additional-column

/* config:
// #additional-column-config
akka.persistence.r2dbc.journal {
  additional-columns {
    "BlogPost" = ["docs.home.journal.EventTitleColumn"]
  }
  custom-table {
    "BlogPost" = event_journal_blogpost
  }
}
// #additional-column-config
 */

// #additional-column
class EventTitleColumn extends AdditionalColumn[BlogPost.Event, String] {

  override val columnName: String = "title"

  override def bind(insert: AdditionalColumn.Insert[BlogPost.Event]): AdditionalColumn.Binding[String] =
    insert.value match {
      case BlogPost.PostAdded(_, title, _) => AdditionalColumn.BindValue(title)
      case _                               => AdditionalColumn.BindNull
    }
}
// #additional-column

File: docs/src/test/java/jdocs/home/journal/EventTitleColumn.java
/*
 * Copyright (C) 2022 - 2025 Lightbend Inc. <https://akka.io>
 */

package jdocs.home.journal;

// #additional-column
import akka.persistence.r2dbc.journal.javadsl.AdditionalColumn;

public class EventTitleColumn extends AdditionalColumn<BlogPost.Event, String> {
  @Override
  public Class<String> fieldClass() {
    return String.class;
  }

  @Override
  public String columnName() {
    return "title";
  }

  @Override
  public Binding<String> bind(Insert<BlogPost.Event> insert) {
    BlogPost.Event event = insert.value();
    if (event instanceof BlogPost.PostAdded) {
      BlogPost.PostAdded added = (BlogPost.PostAdded) event;
      return AdditionalColumn.bindValue(added.title);
    } else {
      return AdditionalColumn.bindNull();
    }
  }
}
// #additional-column
